### PR TITLE
Forward declare execution and memory spaces unconditionally

### PR DESCRIPTION
**Pull request by masterleinad**
This pull request revives https://github.com/kokkos/kokkos/pull/3603. I remember us agreeing that we want to have the execution and memory space for the "usual" backends always defined. Also, we had a recent question on the mailing list asking for it.
In a separate commit, this pull request removes some of the preprocessor guards that were previously necessary.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by DavidPoliakoff** [comment]
@masterleinad , what error message do I get if I do a parallel_for with Cuda without Cuda enabled now? "Use of incomplete type?"

**Comment by masterleinad** [comment]
> @masterleinad , what error message do I get if I do a parallel_for with Cuda without Cuda enabled now? "Use of incomplete type?"

```
In file included from /tmp/kokkos/core/src/Kokkos_Parallel.hpp:62:0,
                 from /tmp/kokkos/core/src/Kokkos_OpenMP.hpp:62,
                 from /tmp/kokkos/core/src/decl/Kokkos_Declare_OPENMP.hpp:49,
                 from /tmp/kokkos/example/build_cmake_in_tree/build/kokkos/KokkosCore_Config_DeclareBackend.hpp:47,
                 from /tmp/kokkos/core/src/Kokkos_Core.hpp:56,
                 from /tmp/kokkos/example/build_cmake_in_tree/cmake_example.cpp:45:
/tmp/kokkos/core/src/impl/Kokkos_FunctorAdapter.hpp: In instantiation of ‘struct Kokkos::Impl::FunctorValueTraits<main(int, char**)::<lambda(long int, long int&)>, Kokkos::Cuda, false>’:
/tmp/kokkos/core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:274:79:   required from ‘class Kokkos::Impl::ParallelReduce<main(int, char**)::<lambda(long int, long int&)>, Kokkos::RangePolicy<Kokkos::Cuda>, Kokkos::InvalidType, Kokkos::OpenMP>’
/tmp/kokkos/core/src/Kokkos_Parallel_Reduce.hpp:873:9:   required from ‘static void Kokkos::Impl::ParallelReduceAdaptor<PolicyType, FunctorType, ReturnType>::execute(const string&, const PolicyType&, const FunctorType&, ReturnType&) [with PolicyType = Kokkos::RangePolicy<Kokkos::Cuda>; FunctorType = main(int, char**)::<lambda(long int, long int&)>; ReturnType = long int; std::__cxx11::string = std::__cxx11::basic_string<char>]’
/tmp/kokkos/core/src/Kokkos_Parallel_Reduce.hpp:1018:76:   required from ‘typename std::enable_if<Kokkos::is_execution_policy<ExecPolicy>::value>::type Kokkos::parallel_reduce(const PolicyType&, const FunctorType&, ReturnType&) [with PolicyType = Kokkos::RangePolicy<Kokkos::Cuda>; FunctorType = main(int, char**)::<lambda(long int, long int&)>; ReturnType = long int; typename std::enable_if<Kokkos::is_execution_policy<ExecPolicy>::value>::type = void]’
/tmp/kokkos/example/build_cmake_in_tree/cmake_example.cpp:69:12:   required from here
/tmp/kokkos/core/src/impl/Kokkos_FunctorAdapter.hpp:1326:35: error: invalid use of incomplete type ‘using tag_type = using type = class Kokkos::Cuda {aka class Kokkos::Cuda}’
       decltype(deduce_reduce_type(tag_type(), &FunctorType::operator()));
                                   ^~~~~~~~~~
In file included from /tmp/kokkos/core/src/Kokkos_Core.hpp:51:0,
                 from /tmp/kokkos/example/build_cmake_in_tree/cmake_example.cpp:45:
/tmp/kokkos/core/src/Kokkos_Core_fwd.hpp:363:7: note: forward declaration of ‘using tag_type = using type = class Kokkos::Cuda {aka class Kokkos::Cuda}’
 class Cuda;
       ^~~~
[...]
```

**Comment by DavidPoliakoff** [comment]
@masterleinad thanks. That's not terrible. I'd love it if we could get the error to be where the user writes Kokkos::Cuda(), but this might make sense? We should discuss this one in the meeting

**Comment by masterleinad** [comment]
Previously,
```
/tmp/kokkos/example/build_cmake_in_tree/cmake_example.cpp: In function ‘int main(int, char**)’:
/tmp/kokkos/example/build_cmake_in_tree/cmake_example.cpp:68:35: error: ‘Cuda’ is not a member of ‘Kokkos’
       Kokkos::RangePolicy<Kokkos::Cuda>(0, n), KOKKOS_LAMBDA(const long i, long& lcount) { lcount += (i % 2) == 0; },
                                   ^~~~
/tmp/kokkos/example/build_cmake_in_tree/cmake_example.cpp:68:35: error: ‘Cuda’ is not a member of ‘Kokkos’
/tmp/kokkos/example/build_cmake_in_tree/cmake_example.cpp:68:39: error: template argument 1 is invalid
       Kokkos::RangePolicy<Kokkos::Cuda>(0, n), KOKKOS_LAMBDA(const long i, long& lcount) { lcount += (i % 2) == 0; },                                      ^
```

**Review by DavidPoliakoff — APPROVED** [review]
Thought I had approved this after our last exchange. I think this will help people quite a bit, and I don't see a breaking path

File: core/src/Kokkos_Core_fwd.hpp
Comment (nliber) [inline core/src/Kokkos_Core_fwd.hpp:333]: Just curious (not a request to change):  why default to available and explicitly mark those that are unavailable?

Also, based on its name it looks like it could be applied to any type (not just execution and memory spaces).  Would `IsSpaceAvailable` be a better name?

File: core/src/Kokkos_Core_fwd.hpp
Comment (nliber) [inline core/src/Kokkos_Core_fwd.hpp:376]: Minor: we should be self-consistent here about `#ifndef` vs `#if !defined`.

**Review by nliber — CHANGES_REQUESTED** [review]

File: core/src/Kokkos_Core_fwd.hpp
Comment (masterleinad) [reply core/src/Kokkos_Core_fwd.hpp:333]: I just took this from @jrmadsen's pull requests and it's currently unused. I'm open to remove this and add something later if desired.

File: containers/unit_tests/TestDynViewAPI.hpp
Comment (dhollman) [inline containers/unit_tests/TestDynViewAPI.hpp:1405]: Non-blocking comment: won't this be a problem for, e.g., HIP also? Shouldn't the `data()` method be SFINAE'd out if this isn't an allowed operation?

**Comment by nmm0** [comment]
Would there be a way to have a typelist of available execution and memory spaces? This might be pretty useful for some of the generic design of #3841 that's used for resilient code.

File: core/src/Kokkos_Core_fwd.hpp
Comment (jrmadsen) [reply core/src/Kokkos_Core_fwd.hpp:333]: @nliber I have used that style in timemory bc it ensures that someone cannot specialize something to be turned on that there isn't support for in the build system, i.e. explicit specialization on the types that definitely should be turned off. Although this is probably less of a concern in Kokkos than it is in timemory.

**Comment by masterleinad** [comment]
> Would there be a way to have a typelist of available execution and memory spaces? This might be pretty useful for some of the generic design of #3841 that's used for resilient code.

That shouldn't be too hard but I would discuss that in a separate pull request once/if we agree on wanting this one. https://github.com/kokkos/kokkos/pull/3623 would also expose something like this.

**Review by nliber — APPROVED** [review]

**Comment by masterleinad** [comment]
Retest this please.

**Review by dhollman — APPROVED** [review]
LGTM. Some minor changes that would be nice if we have time.

File: containers/unit_tests/TestDynViewAPI.hpp
Comment (masterleinad) [reply containers/unit_tests/TestDynViewAPI.hpp:1405]: I created an issue for that: https://github.com/kokkos/kokkos/issues/3894

File: core/src/Kokkos_Core_fwd.hpp
Comment (masterleinad) [reply core/src/Kokkos_Core_fwd.hpp:339]: Done.

**Comment by masterleinad** [comment]
I addressed the review comments and also implemented a type list in the last commit. I am happy to split and discuss somewhere else if desired.

**Review by DavidPoliakoff — APPROVED** [review]

**Review by dalg24 — CHANGES_REQUESTED** [review]
Make sure you do not break the plugins interface.

Neither you nor #3603 discuss thoroughly how the changes proposed will impact user code.  What will they be able to do that they could not before and what are the limitations.

For instance will your PR help with
https://github.com/arborx/ArborX/blob/ae19f4a44d7a60da13b1ee86346724b6314384ff/src/ArborX_LinearBVH.hpp#L93-L102
What about
https://github.com/arborx/ArborX/blob/ae19f4a44d7a60da13b1ee86346724b6314384ff/benchmarks/bvh_driver/bvh_driver.cpp#L100-L106

File: core/src/Kokkos_Concepts.hpp
Comment (Rombur) [reply core/src/Kokkos_Concepts.hpp:337]: Yeah it looks like it's missing :/

**Comment by masterleinad** [comment]
> Make sure you do not break the plugins interface.

I just wanted to see if it is feasible to at least remove the now redundant files and I am happy to remove the changes in the last few commits and discuss somewhere else.

> Neither you nor #3603 discuss thoroughly how the changes proposed will impact user code. What will they be able to do that they could not before and what are the limitations.

The changes are backward-compatible since we only introduce a few new forward declarations (and a type list with the available memory and execution spaces; as I said early we can also split this up). Multiple people/projects have asked for the type lists (including `ArborX`).
Further advantages are that users can remove header guards as long as they really only use a class declaration for guarding code (as demonstrated in the test suite). This plays nicely with `if constexpr`.
Drawbacks are the inferior error messages as discussed in @DavidPoliakoff's question in this pull request.

> For instance will your PR help with
> [arborx/ArborX@`ae19f4a`/src/ArborX_LinearBVH.hpp#L93-L102](https://github.com/arborx/ArborX/blob/ae19f4a44d7a60da13b1ee86346724b6314384ff/src/ArborX_LinearBVH.hpp#L93-L102)

Yes, we can remove the preprocessor guards.

> What about
> [arborx/ArborX@`ae19f4a`/benchmarks/bvh_driver/bvh_driver.cpp#L100-L106](https://github.com/arborx/ArborX/blob/ae19f4a44d7a60da13b1ee86346724b6314384ff/benchmarks/bvh_driver/bvh_driver.cpp#L100-L106)

The preprocessor conditional can be replaced by `if constexpr(IsSpaceAvailable<Cuda>::value)` (is we were to enforce C++17) but I don't see a huge benefit here.

File: core/src/fwd/Kokkos_Fwd_SYCL.hpp
Comment (masterleinad) [reply core/src/fwd/Kokkos_Fwd_SYCL.hpp:51]: Yes

File: core/src/KokkosCore_Config_FwdBackend.hpp
Comment (masterleinad) [reply core/src/KokkosCore_Config_FwdBackend.hpp:60]: We can of course keep the separate files instead (and modify them accordingly) but we still need to always include them if we want this feature.

**Comment by dalg24** [comment]
The type list is a separate concern.  We could provide it regardless of whether we introduce forward declarations or not.

**Comment by masterleinad** [comment]
> The type list is a separate concern. We could provide it regardless of whether we introduce forward declarations or not.

Yes, that's what I'm saying. This pull request makes it easy to add it (for the "standard" backends; I'm not quite sure how/if this would expand easily to external backends using the plugin system).

Apparently, we have a lot of approvals here and people seem to be in favor of the general idea of forward declaring the backends. What would be an agreeable version? 
Keeping all the forward declaration in `Kokkos_Core_fwd.hpp` and the duplicates created for the plugin system is basically [`c2ebddf` (#3838)](https://github.com/kokkos/kokkos/pull/3838/commits/c2ebddffae0e6b5dbb31d2bd4bf62dfeb09a477b). I am fine with removing the type lists if we want to discuss this somewhere else.

We could also combine the new forward declarations with the ones by the plugin system (or do this cleanup in a follow-up pull request).

**Comment by masterleinad** [comment]
I rolled back to a version not interfering with the plugin system.

File: core/src/Kokkos_Concepts.hpp
Comment (crtrott) [reply core/src/Kokkos_Concepts.hpp:337]: Are we addressing this?

File: core/unit_test/TestTeamBasic.hpp
Comment (crtrott) [inline core/unit_test/TestTeamBasic.hpp:69]: I don't like this change, since it will look like the test is actually running for OpenMPTarget while it actually isn't.

File: core/src/Kokkos_Core_fwd.hpp
Comment (crtrott) [inline core/src/Kokkos_Core_fwd.hpp:345]: this Serial still doesn't seem to be moved out of here?

File: core/src/Kokkos_Core_fwd.hpp
Comment (crtrott) [inline core/src/Kokkos_Core_fwd.hpp:458]: What are those two things used for?

**Review by crtrott — CHANGES_REQUESTED** [review]

File: core/src/Kokkos_Concepts.hpp
Comment (masterleinad) [reply core/src/Kokkos_Concepts.hpp:337]: 62fd5f647c

Comment (masterleinad) [reply core/src/Kokkos_Concepts.hpp:341]: Updated in [`0739f65` (#3838)](https://github.com/kokkos/kokkos/pull/3838/commits/0739f656ca13dcacb794be1dded316c6d64795fb).

File: core/src/Kokkos_Core_fwd.hpp
Comment (masterleinad) [reply core/src/Kokkos_Core_fwd.hpp:345]: Yes, you're right. I want to combine this with the `fwd/Kokkos_FWD*` files anyway in the next commit.

File: core/src/Kokkos_Core_fwd.hpp
Comment (masterleinad) [reply core/src/Kokkos_Core_fwd.hpp:434]: Several people were asking for an iterable list of active execution and memory spaces and that is what these two variables do. We can add that in a follow-up pull request, though.

**Comment by jrmadsen** [comment]
Couple thoughts here:

In `Kokkos/Kokkos_utilities.hpp`, I think we should add:

```cpp
template <typename Tp, typename Types>
struct is_one_of {
  template <class...>
  struct is_one_of_impl {
    static constexpr bool value = false;
  };

  template <class F, class S, template <class...> class Tuple, class... T>
  struct is_one_of_impl<F, S, Tuple<T...>> {
    static constexpr bool value =
        std::is_same<F, S>::value || is_one_of_impl<F, Tuple<T...>>::value;
  };

  template <class F, class S, template <class...> class Tuple, class... T>
  struct is_one_of_impl<F, Tuple<S, T...>> {
    static constexpr bool value = is_one_of_impl<F, S, Tuple<T...>>::value;
  };

public:
    static constexpr bool value = is_one_of_impl<Tp, Types>::value;
    using type = std::conditional_t<value, std::true_type, std::false_type>;
};
```

So that things like:

```cpp
  using do_not_use_host_memory_space = std::conditional_t<
      std::is_same<memory_space, Kokkos::HostSpace>::value ||
          std::is_same<memory_space, Kokkos::CudaUVMSpace>::value ||
          std::is_same<memory_space, Kokkos::CudaHostPinnedSpace>::value ||
          std::is_same<memory_space,
                       Kokkos::Experimental::HIPHostPinnedSpace>::value ||
          std::is_same<memory_space,
                       Kokkos::Experimental::SYCLSharedUSMSpace>::value,
      memory_space, Kokkos::HostSpace>;
```

(i.e. multiple `std::is_same<...> || ...`) can be much more cleanly expressed via:

```cpp
  using do_not_use_host_memory_space = 
      conditional_t<is_one_of<memory_space, 
                              type_list<Kokkos::HostSpace, 
                                              Kokkos::CudaUVMSpace, 
                                              Kokkos::CudaHostPinnedSpace, 
                                              Kokkos::Experimental::HIPHostPinnedSpace,
                                              Kokkos::Experimental::SYCLSharedUSMSpace>>::value,
          memory_space, Kokkos::HostSpace>;
```

Along those lines, I would like for us to have enumerations which map to the spaces and enable looking up string IDs for the available memory and execution spaces. This would benefit pykokkos-base (which needs this) and help logging/tools (i.e. print/log the spaces that available).

**Comment by masterleinad** [comment]
I think the most concerns are still about error messages and the scope of this pull request.
Since there seems to be increasing demand, it would probably make sense to provide a minimal solution (only forward declarations) here and everything else (@jrmadsen's suggestions, `IsSpaceAvailable` and similar) in a follow-up pull request.

**Comment by jrmadsen** [comment]
What error messages?

**Comment by masterleinad** [comment]
> What error messages?

Error messages when using spaces that only have forward declarations, basically https://github.com/kokkos/kokkos/pull/3838#issuecomment-795469669.

**Comment by jrmadsen** [comment]
So the issue is that what was previously an error is now just a different error? Let's just add in static asserts and make an entry in the wiki about what an undefined space means

**Comment by masterleinad** [comment]
Something like
```diff
diff --git a/core/src/Kokkos_View.hpp b/core/src/Kokkos_View.hpp
index 117cb3b05..2eed6f12b 100644
--- a/core/src/Kokkos_View.hpp
+++ b/core/src/Kokkos_View.hpp
@@ -208,6 +208,8 @@ struct ViewTraits<
     typename std::enable_if<Kokkos::Impl::is_space<Space>::value>::type, Space,
     Prop...> {
   // Specify Space, memory traits should be the only subsequent argument.
+  static_assert(IsSpaceAvailable<Space>::value,
+                "The requested space has not been enabled!");
 
   static_assert(
       std::is_same<typename ViewTraits<void, Prop...>::execution_space,
diff --git a/core/src/fwd/Kokkos_Fwd_SYCL.hpp b/core/src/fwd/Kokkos_Fwd_SYCL.hpp
index 5dcef3793..03a87a7e8 100644
--- a/core/src/fwd/Kokkos_Fwd_SYCL.hpp
+++ b/core/src/fwd/Kokkos_Fwd_SYCL.hpp
@@ -60,6 +60,26 @@ template <>
 struct IsSpaceAvailable<Experimental::SYCLDeviceUSMSpace> : std::false_type {};
 template <>
 struct IsSpaceAvailable<Experimental::SYCLSharedUSMSpace> : std::false_type {};
+
+template <typename T>
+struct is_execution_space;
+
+template <typename T>
+struct is_space;
+
+template <>
+struct is_execution_space<Kokkos::Experimental::SYCL> : public std::true_type {
+};
+
+template <>
+struct is_space<Kokkos::Experimental::SYCLDeviceUSMSpace>
+    : public std::true_type {};
+
+template <>
+struct is_space<Kokkos::Experimental::SYCLSharedUSMSpace>
+    : public std::true_type {};
+
 #endif
+
 }  // namespace Kokkos
 #endif
diff --git a/core/src/traits/Kokkos_ExecutionSpaceTrait.hpp b/core/src/traits/Kokkos_ExecutionSpaceTrait.hpp
index 3a8823e94..8f5b29f97 100644
--- a/core/src/traits/Kokkos_ExecutionSpaceTrait.hpp
+++ b/core/src/traits/Kokkos_ExecutionSpaceTrait.hpp
@@ -88,6 +88,8 @@ struct ExecutionSpaceTrait : TraitSpecificationBase<ExecutionSpaceTrait> {
     static constexpr auto execution_space_is_defaulted = false;
 
     using execution_space = ExecSpace;
+    static_assert(IsSpaceAvailable<ExecSpace>::value,
+                  "The requested execution space has not been enabled!");
   };
 };
```
gives better error messages like
```
In file included from /tmp/kokkos_new/core/src/impl/Kokkos_AnalyzePolicy.hpp:53:0,
                 from /tmp/kokkos_new/core/src/Kokkos_ExecPolicy.hpp:52,
                 from /tmp/kokkos_new/core/src/Kokkos_Parallel.hpp:53,
                 from /tmp/kokkos_new/core/src/Kokkos_Serial.hpp:57,
                 from /tmp/kokkos_new/core/src/decl/Kokkos_Declare_SERIAL.hpp:49,
                 from /tmp/kokkos_new/example/build_cmake_in_tree/build_ne/kokkos/KokkosCore_Config_DeclareBackend.hpp:47,
                 from /tmp/kokkos_new/core/src/Kokkos_Core.hpp:56,
                 from /tmp/kokkos_new/example/build_cmake_in_tree/cmake_example.cpp:45:
/tmp/kokkos_new/core/src/traits/Kokkos_ExecutionSpaceTrait.hpp: In instantiation of ‘struct Kokkos::Impl::ExecutionSpaceTrait::mixin_matching_trait<Kokkos::Experimental::SYCL, Kokkos::Impl::AnalyzeExecPolicy<void> >’:
/tmp/kokkos_new/core/src/impl/Kokkos_AnalyzePolicy.hpp:159:8:   required from ‘struct Kokkos::Impl::AnalyzeExecPolicyUseMatcher<void, Kokkos::Impl::type_list<Kokkos::Impl::ExecutionSpaceTrait, Kokkos::Impl::IndexTypeTrait, Kokkos::Impl::ScheduleTrait, Kokkos::Impl::IterationPatternTrait, Kokkos::Impl::WorkItemPropertyTrait, Kokkos::Impl::LaunchBoundsTrait, Kokkos::Impl::OccupancyControlTrait, Kokkos::Impl::GraphKernelTrait, Kokkos::Impl::WorkTagTrait>, Kokkos::Experimental::SYCL>’
/tmp/kokkos_new/core/src/impl/Kokkos_AnalyzePolicy.hpp:106:8:   required from ‘struct Kokkos::Impl::AnalyzeExecPolicy<void, Kokkos::Experimental::SYCL>’
/tmp/kokkos_new/core/src/impl/Kokkos_AnalyzePolicy.hpp:209:8:   required from ‘struct Kokkos::Impl::ExecPolicyTraitsWithDefaults<Kokkos::Impl::AnalyzeExecPolicy<void, Kokkos::Experimental::SYCL> >’
/tmp/kokkos_new/core/src/impl/Kokkos_AnalyzePolicy.hpp:224:8:   required from ‘struct Kokkos::Impl::PolicyTraits<Kokkos::Experimental::SYCL>’
/tmp/kokkos_new/core/src/Kokkos_ExecPolicy.hpp:91:7:   required from ‘class Kokkos::RangePolicy<Kokkos::Experimental::SYCL>’
/tmp/kokkos_new/example/build_cmake_in_tree/cmake_example.cpp:72:59:   required from here 
/tmp/kokkos_new/core/src/traits/Kokkos_ExecutionSpaceTrait.hpp:91:5: error: static assertion failed: The requested execution space has not been enabled!
     static_assert(IsSpaceAvailable<ExecSpace>::value,
     ^~~~~~~~~~~~~
```
I am happy to add that for all spaces if desired.

**Pull request closed without merging**
